### ISSUE TITLE
PHP intelephense lsp setting not working!

### ISSUE THREAD

**Issue by Ok9xNirab** [zed-industries/zed#4258]
According to [intelephense  docs](https://github.com/bmewburn/intelephense-docs/blob/master/gettingStarted.md#environment), i want to set `wordpress` stub inside `intelephense.stubs`

So, i just configure the editor based on zed docs : 

```json
  "lsp": {
    "intelephense": {
      "initialization_options": {
        "stubs": [
          "apache",
          "bcmath",
          "bz2",
          "calendar",
          "com_dotnet",
          "Core",
          "ctype",
          "curl",
          "date",
          "dba",
          "dom",
          "enchant",
          "exif",
          "FFI",
          "fileinfo",
          "filter",
          "fpm",
          "ftp",
          "gd",
          "gettext",
          "gmp",
          "hash",
          "iconv",
          "imap",
          "intl",
          "json",
          "ldap",
          "libxml",
          "mbstring",
          "meta",
          "mysqli",
          "oci8",
          "odbc",
          "openssl",
          "pcntl",
          "pcre",
          "PDO",
          "pdo_ibm",
          "pdo_mysql",
          "pdo_pgsql",
          "pdo_sqlite",
          "pgsql",
          "Phar",
          "posix",
          "pspell",
          "random",
          "readline",
          "Reflection",
          "session",
          "shmop",
          "SimpleXML",
          "snmp",
          "soap",
          "sockets",
          "sodium",
          "SPL",
          "sqlite3",
          "standard",
          "superglobals",
          "sysvmsg",
          "sysvsem",
          "sysvshm",
          "tidy",
          "tokenizer",
          "xml",
          "xmlreader",
          "xmlrpc",
          "xmlwriter",
          "xsl",
          "Zend OPcache",
          "zip",
          "zlib",
          "wordpress"
        ]
      }
    }
  }
```

But didn't working !!

<img width="1552" alt="Screenshot 2024-01-21 at 18 53 19" src="https://github.com/zed-industries/community/assets/40653908/9e60b9ba-fc48-4000-a023-7e6432535044">

**Thanks**

**Comment by lukeholder** [comment]
I would love good PHP support in Zed. People don't realise how much PHP is used.

**Comment by basitcodeenv** [comment]
I need support for PHP in Zed

**Comment by Tsolmonx** [comment]
Does anyone know zed supports intelephense premium ?

**Comment by tjohnman** [comment]
This is important for me, too. Is it a bug, or are we supposed to set the configuration in some other way?

**Comment by sbusch** [comment]
Premium should work, although not tested by myself: https://github.com/zed-industries/zed/issues/6965

**Comment by dinhtungdu** [comment]
I have a similar issue, `includePaths` is also not working for me. Here is the config I'm using:

```
  "lsp": {
    "intelephense": {
      "initialization_options": {
        "environment": {
          "includePaths": ["/path/to/lib"]
        }
      }
    }
  }
```

**Comment by ggarciajr** [comment]
braces format is also not being respected

**Comment by usizu** [comment]
Has anyone gotten Wordpress (and WooCommerce / ACF) stubs working in Zed?

**Comment by JeremyOB** [comment]
1+

**Comment by pjv** [comment]
This is a horrible kludge, but I got intelephense to respect the `includePaths` by directly editing one of the intelephense_data files in a private directory somewhere.

First I opened the zed log (`zed: open log`) and I scanned through until I saw something like this:

`2024-06-12T04:09:28-05:00 [INFO] open paths ["/private/var/folders/jk/h5pbftgn6nj_7w6rdqkqlg2w0000gq/T/intelephense/51d78cbd/intelephense_data_1"]`

That file looked like this:

`{"timestamp":1718183316791,"counter":{"next":1296,"holes":[]},"environment":{"documentRoot":"file:///<my wp project>“,"includePaths":[],"phpVersion":"8.3.0","shortOpenTag":false},"intelephenseVersion":"1.10.4"}`

I added a couple paths to `includePaths`:

`{"timestamp":1718183316791,"counter":{"next":1296,"holes":[]},"environment":{"documentRoot":"file:///<my wp project>“,"includePaths":["/path/to/wordpress","/path/to/woocommerce"],"phpVersion":"8.3.0","shortOpenTag":false},"intelephenseVersion":"1.10.4"}`

…then restarted zed and opened some php files with a bunch of wordpress function calls and no error messages from intelephense. Go to definition does not work, but mousing over wp function calls does show me the function’s template.

obviously there is some kind of disconnect between defining `lsp.intelephense.initialization_options` in your zed settings and those initialization_options getting propagated to intelephense.

**Comment by plugd-in** [comment]
I got LSP settings working for PHP, since I needed the stub for ssh2. Basically, I had to update the `zed_extension_api` for PHP to 0.0.6, and then add  a `fn language_server_workspace_configuration(...)` implementation which I basically grabbed from one of the other extensions. I had to build the extension using `zed/crates/extension_cli`, then I bumped the PHP extension version I built so it wouldn't be replaced, and copied the built extension to `.local/share/zed/extensions/installed`, I believe.

**Comment by federicoangelli** [comment]
> I got LSP settings working for PHP, since I needed the stub for ssh2. Basically, I had to update the `zed_extension_api` for PHP to 0.0.6, and then add a `fn language_server_workspace_configuration(...)` implementation which I basically grabbed from one of the other extensions. I had to build the extension using `zed/crates/extension_cli`, then I bumped the PHP extension version I built so it wouldn't be replaced, and copied the built extension to `.local/share/zed/extensions/installed`, I believe. I forked Zed and added my commit to my fork. After updating my PHP extension, I simply specified the stubs like so: ![image](https://private-user-images.githubusercontent.com/84760193/345958870-7770f2cd-9fe6-4eab-b3f5-a2d1ce040dae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA5NjcxMjcsIm5iZiI6MTcyMDk2NjgyNywicGF0aCI6Ii84NDc2MDE5My8zNDU5NTg4NzAtNzc3MGYyY2QtOWZlNi00ZWFiLWIzZjUtYTJkMWNlMDQwZGFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE0VDE0MjAyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzYzRiODUwNWM1NWNiNDcxZmMwZmE1NjAxNDEyYWViOTljYTBmMmU5NTEzYWU1OWFlYTlhNTY2ZTEyNWJhNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QvlRa1bjD9f_ySviHOKyeHQtAkTEjG_eay8h8-DH8Qo)
> 
> If anyone wants to take over from here, feel free to fork Zed, cherry-pick my commit and then create a PR.

Could you please give a more detailed step-by-step explanation of what you had to do? I've forked Zed, installed Rust and Xcode (per [these instructions](https://github.com/zed-industries/zed/blob/main/docs/src/development/macos.md)), modified the files as in your commit, ran `cargo build --release`, got that "target" directory, but I'm not sure what I should do next. I don't have the `.local/share/zed` directory. Also, can you build _just_ the extension? It seems that I have built the whole Zed app using that command. And how do I "bump" the extension's version? Thanks! 🙏

**Comment by plugd-in** [comment]
@federicoangelli Yeah, no problem. Basically, part of the release process, I assume is to build the extensions into a WASM plugin, and those plugins are downloaded from some plugin/release server. I think if you run Zed and install the PHP extension, you should have a `~/.local/share/zed/extensions/installed/php/` directory where the downloaded extension is placed.

Now, the tool that's actually used to build the extension (which was somewhat hard to figure out), is in `crates/extension_cli`. You can build it with `cargo build -p extension_cli --release`. After building the it, you can go on to build your extension like `mkdir /tmp/zed_build && ./target/release/zed-extension --source-dir extensions/php` --output-dir . --scratch-dir /tmp/zed_build`. For details on the arguments, you can also use `./target/release/zed-extension --help`. You should end up with a file, `archive.tar.gz`, with the extension in it. You can unpack it with `tar -xzvf archive.tar.gz`, and your extension should look a lot like what was present in `~/.local/share/zed/extensions/installed/php/`.

You should then edit the manifest.json file to have some version greater than `0.0.2`, such as `0.0.3` or higher, which will prevent Zed from automatically replacing your extension with the version with lesser LSP setting support.

Then, you move your built extension into the `~/.local/share/zed/extensions/installed/php/` directory, replacing the files that were currently there.

Do note that I did all this in Linux, and the case might be different for Mac, although the biggest difference might be where the extensions are stored.

**Comment by federicoangelli** [comment]
@plugd-in You're right, I've managed to make it work following your steps. On macos, the path to where extensions are installed is `~/Library/Application\ Support/Zed/extensions/installed/`, that's the only difference in our context. Thanks! 🙏

**Comment by maxdeviant** [comment]
v0.1.2 of the PHP extension now has support for reading the LSP settings from the Zed settings:

Here's how the settings should be structured in your `settings.json`:

```json
{
  "lsp": {
    "intelephense": {
      "settings": {
        "stubs": ["foo", "bar"],
        "files": {
          "maxSize": 1000001
        }
      }
    }
  }
}
```

**Comment by tiojoca** [comment]
It seems to work but I still get errors from code that is recognized:
<img width="639" alt="image" src="https://github.com/user-attachments/assets/e4331156-d676-4e5c-9dca-59d16065b42e">

**Comment by charmfocus** [comment]
> I have a similar issue, `includePaths` is also not working for me. Here is the config I'm using:
> 
> ```
>   "lsp": {
>     "intelephense": {
>       "initialization_options": {
>         "environment": {
>           "includePaths": ["/path/to/lib"]
>         }
>       }
>     }
>   }
> ```

https://github.com/zed-industries/zed/issues/16009#issuecomment-2278515131

**Comment by nadavmoreno** [comment]
this one is working for me:
```
  "lsp": {
    "intelephense": {
      "settings": {
        "environment": {
          "includePaths": ["/path/to/lib"]
        },
        "stubs": [
          "wordpress",
          "woocommerce",
          "acf-pro",
          "wordpress-globals",
          "wp-cli"
        ],
        "files": {
          "maxSize": 5000000,
          "excluded": ["**/node_modules/**", "**/vendor/**"]
        }
      }
    }
 }
```

**Comment by datgausaigon** [comment]
I have opened an issue:
https://github.com/zed-industries/zed/issues/17236
And I found this issue of yours related to it. Please join me in supporting it.
Thanks 
🐻

**Comment by vuon9** [comment]
Hi Zed team,

TBH, nested JSON is so ugly in this specific case, and it will also make configuring works become more difficult and I don't want to buy a bigger monitor just because of switching to a new editor. Could be  great if it be the same as VSCode?

![image](https://github.com/user-attachments/assets/3e8935ec-4c22-42d2-80d0-6176e8e186bb)

**Comment by admtech** [comment]
```
"lsp": {
    "intelephense": {
      "settings": {
        "environment": {
          "includePaths": [
            "/home/test/projects/framework",
            "/home/test/projects/core"
          ]
        },
        "stubs": ["openswoole"],
        "files": {
          "excluded": [
            "**/node_modules/**",
            "**/vendor/**",
            "**/old/**",
            "**/_old/**"
          ]
        }
      }
    }
  }

```
Unfortunately, this does **not work for me**. I just want to remove the directories /old/ and _old from code completion. But the setting is simply ignored and classes from _old are displayed. Also includePaths and stubs do not work.
PHP extension v0.2.0

Too bad, I would have loved to switch to the zed IDE for PHP development.

**Comment by iansvo** [comment]
> v0.1.2 of the PHP extension now has support for reading the LSP settings from the Zed settings:
> 
> Here's how the settings should be structured in your `settings.json`:
> 
> {
>   "lsp": {
>     "intelephense": {
>       "settings": {
>         "stubs": ["foo", "bar"],
>         "files": {
>           "maxSize": 1000001
>         }
>       }
>     }
>   }
> }

This ended up being what I needed, thanks for sharing.

**Comment by notpeter** [comment]
I'm going to move this to the [Zed Language support discussions](https://github.com/zed-industries/zed/discussions/categories/language-support).

If folks have suggestions for incorporate additional configuration suggestions to always happy to PRs to [docs/languages/php.md](https://github.com/zed-industries/zed/blob/main/docs/src/languages/php.md?plain=1).

Thanks!
